src/main/java/de/tum/in/www1/artemis/config/migration/entries/MigrationEntry20230808_203400.java

```diff
@@ -0,0 +1,460 @@
+            errorMap.put(exercise, true);
+        }
+        try {
+            ciMigrationService.orElseThrow().overrideBuildPlanRepository(exercise.getTemplateBuildPlanId(), "tests", exercise.getTestRepositoryUrl(), exercise.getBranch());
+        }
+        catch (VersionControlException | ContinuousIntegrationException | NoSuchElementException | RestClientException e) {
+            log.warn("Failed to replace tests repository in template build plan for exercise {} with buildPlanId {}", exercise.getId(), exercise.getTemplateBuildPlanId(), e);
+            errorMap.put(exercise, true);
+        }
+        for (var auxiliary : auxiliaryRepositories) {
+            try {
+                ciMigrationService.orElseThrow().overrideBuildPlanRepository(exercise.getTemplateBuildPlanId(), auxiliary.getName(), auxiliary.getRepositoryUrl(),
+                        exercise.getBranch());
+            }
+            catch (VersionControlException | ContinuousIntegrationException | NoSuchElementException | RestClientException e) {
+                log.warn("Failed to replace auxiliary repository {} in template build plan for exercise {} with buildPlanId {}", auxiliary.getName(), exercise.getId(),
+                        exercise.getTemplateBuildPlanId(), e);
+                errorMap.put(exercise, true);
+            }
+        }
+        try {
+            ciMigrationService.orElseThrow().overrideRepositoriesToCheckout(exercise.getTemplateBuildPlanId(), auxiliaryRepositories);
+        }
+        catch (VersionControlException | ContinuousIntegrationException | NoSuchElementException | RestClientException e) {
+            log.error("Failed to replace repositories in template build plan for exercise {} with buildPlanId {}", exercise.getId(), exercise.getTemplateBuildPlanId(), e);
+            errorMap.put(exercise, true);
+        }
+    }
+
+    private void evaluateErrorMap(Map<ProgrammingExercise, Boolean> errorMap) {
+        if (errorMap.isEmpty()) {
+            return;
+        }
+
+        List<Long> failedExerciseIdsForTemplates = errorMap.entrySet().stream().filter(entry -> Boolean.TRUE.equals(entry.getValue())).map(entry -> entry.getKey().getId())
+                .toList();
+        List<Long> failedExerciseIdsForSolutions = errorMap.entrySet().stream().filter(entry -> Boolean.FALSE.equals(entry.getValue())).map(entry -> entry.getKey().getId())
+                .toList();
+        log.error("Failed to migrate {} programming exercises", errorMap.size());
+        log.error("Failed to migrate template build plan for exercises: {}", failedExerciseIdsForTemplates);
+        log.error("Failed to migrate solution build plan for exercises: {}", failedExerciseIdsForSolutions);
+        log.warn("Please check the logs for more information. Then either fix the issues and rerun the migration or "
+                + "fix the build plans yourself and mark the migration as run.");
```

Comment:
IMO there should be some kind of rollback for when the migration completely or mostly fails. We are dealing with pretty huge, complicated, important data here 🤔

Reply:
A rollback is not feasible, the rollback would take as long as the migration and then we still could run into errors on every single API call.